File: PrimeCalc.py
# I didn't use math library this time

'''
Class Name: PrimeC
This class is for the prime pair calculation.
'''
class PrimeC:
    def __init__(self):
        self.gap = -1
        self.determineNum = 2
        self.pairNum = 0
        self.pps = [ ]
        self.rangeNum = -1

    '''
    Function Name: setGap
    Parameter: None
    Return Value: None
    set gap variable
    '''
    def setGap(self):
        self.gap = input("Enter a gap: ")

    '''
    Function Name: setPairNum
    Parameter: None
    Return Value: None
    set the number of the prime numbers in the pair
    '''
    def setPairNum(self):
        self.pairNum = input("Enter # of prime numbers in the pair: ")

    '''
    Function Name: setRange
    Parameter: None
    Return Value: None
    set the range to calculate
    '''
    def setRange(self):
        self.rangeNum = input("Enter the range to calculate: ")

    '''
    Function Name: setPrivates
    Parameter: None 
    Return Value: None
    set the private variables for the calculation
    '''
    def setPrivates(self):
        self.setGap()
        self.setPairNum()
        self.setRange()

    '''
    Function Name: isPrime
    Parameter: int check
    Return Value: boolean
    check if the check variable is prime
    '''
    def isPrime(self, check):
        for i in range(2, check):
            if check % i == 0:
                return False
        return True

    '''
    Function Name: checkPrimePair
    Parameter: int index
    Return Value: boolean
    check if the index is prime and the previous number(s) with the gap exist(s)
    '''
    def checkPrimePair(self, index):
        return (index - ((self.pairNum - 1) * self.gap) > 1 and self.isPrime(index))

    '''
    Function Name: checkBefores
    Parameter: int index
    Return Value: boolean
    check the previous numbers (with the gap) are prime
    '''
    def checkBefores(self, index):
        index -= self.gap
        for i in range(self.pairNum - 1):
            if not self.isPrime(index):
                return False
            index -= self.gap
        return True

    '''
    Function Name: setPair
    Parameter: int max
    Return Value: None
    set the prime pair into the prime pair list
    '''
    def setPair(self, max):
        pp = [ ]
        for i in range(self.pairNum):
            pp.append(max - (self.pairNum - 1 - i) * self.gap)
        self.pps.append(pp)

    '''
    Function Name: calc
    Parameter: None
    Return Value: None
    caclulate the prime pair list in the range that the user set
    '''
    def calc (self):
        for i in range(self.determineNum, self.rangeNum):
            if self.checkPrimePair(i) and self.checkBefores(i):
                self.setPair(i)

    '''
    Function Name: checkTypePair
    Parameter: None
    Return Value: None
    print out the type of the prime pair
    '''
    def checkTypePair(self):
        if self.gap == 2:
            print("Twin Prime:")
        elif self.gap == 4:
            print("Cousin Prime:")
        elif self.gap == 6:
            print("Sexy Prime:")
        else:
            print("prime numbers with " + str(self.gap) + " gap:")

    '''
    Function Name: displayCount
    Parameter: None
    Return Value: boolean
    check if the prime pair(s) exist(s) in the range
    print out the number of the prime pair 
    '''
    def displayCount(self):
        if len(self.pps) == 0:
            return False
        print("COUNT: ", len(self.pps))
        return True
    '''
    Function Name: display
    Parameter: None
    Return Value: None
    Display the information
    (type of the prime pair, the prime pair, and the number of the pair)
    '''
    def display(self):
        if not self.displayCount():
            print("NO PRIME PAIRS IN ", self.rangeNum)
            return
        self.checkTypePair()
        for i in self.pps:
            print(i)

File: PrimeCalc.cpp
//
//  PrimeCalc.cpp
//  PrimeProject
//
//  Created by Kenny Kim on 9/20/22.
//
#include "PrimeCalc.h"
#include <stdio.h>
#include <iostream>

using namespace std;

/* Constructor */
PrimeCalc::PrimeCalc(){
    gap = -1;
    pairNum = -1;
    rangeNum = -1;
}

/*
 Function Name: setGap
 Parameter: None
 Return Type: None
 set Gap value
 */
void PrimeCalc::setGap(){
    cout << "Enter a gap: ";
    cin >> this->gap;
    cout << endl;
}

/*
 Function Name: setPairNum
 Parameter: None
 Return Value: None
 set the # of prime numbers in the pair
 */
void PrimeCalc::setPairNum(){
    cout << "Enter # of numbers in the pair: ";
    cin >> this->pairNum;
    cout << endl;
}

/*
 Function Name: setRange
 Parameter: None
 Return Value: None
 set the range to calculate the prime pairs
 */
void PrimeCalc::setRange(){
    cout << "Enter the range to calculate: ";
    cin >> this->rangeNum;
    cout << endl;
}

/*
 Function Name: setPairList
 Parameter: None
 Return Value: None
 set the pair list
 */
void PrimeCalc::setPairList(){
    pairListSize = (rangeNum / pairNum);
    pairList = new int*[pairListSize];
    for (int i = 0; i < rangeNum; ++i)
        pairList[i] = new int[pairNum];
}

/* setPrivates(): set private variables */
void PrimeCalc::setPrivates(){
    setGap();
    setPairNum();
    setRange();
    setPairList();
    
}

// I don't use math library for this project.
/*
 Function Name: isPrime
 Parameter: int checkNum
 Return Value: boolean
 Check if the checkNum is prime or not
*/
bool PrimeCalc::isPrime(int checkNum){
    for (int i = 2; i < checkNum; ++i){
        if (checkNum % i == 0)
            return false;
    }
    return true;
}

/*
 Function Name: checkPair
 Parameter: int max
 Return Value: boolean
 set the prime pair into the pair list
 */
bool PrimeCalc::setPair(int max){
    for (int k = 0; k < pairNum; ++k){
        pairList[count][k] = max - (pairNum - 1 - k) * gap;
    }
    ++count;
    return true;
}

/*
 Function Name: checkPrimePair
 Parameter: int index
 Return Value: boolean
 check if the numer has number(s)(depends on pair #) with a specific gap & it's prime
 */
bool PrimeCalc::checkPrimePair(int index){
    return (index - ((pairNum - 1) * gap) > 1 && isPrime(index));
}

/*
 Function Name: checkBefores
 Parameter: int index
 Return Value: boolean
 check if the previous numbers are prime
 */
bool PrimeCalc::checkBefores(int index){
    int* tempPair = new int[pairNum];
    int checkBefore = index - gap;
    for (int i = 0; i < pairNum - 1; ++i){
        if (!isPrime(checkBefore)){
            delete[] tempPair;
            tempPair = nullptr;
            return false;
        }
        tempPair[i] = checkBefore;
        checkBefore -= gap;
    }
    delete[] tempPair;
    tempPair = nullptr;
    return true;
}

/*
 Function Name: calc
 Parameter: None
 Return Value: None
 calculate the prime pair list
 */
void PrimeCalc::calc(){
    for (int i = determineNum; i < rangeNum; ++i){
        if (checkPrimePair(i)){
            if (checkBefores(i))
                setPair(i);
        }
    }
}

/*
 Function Name: display
 Parameter: None
 Return Value: None
 display prime pairs in the specific range that the user set.
 */
void PrimeCalc::display(){
    this->checkTypePair();
    for (int i = 0; i < count; ++i){
        cout << "(";
        for (int j = 0; j < pairNum; ++j){
            cout << pairList[i][j];
            if (j != pairNum - 1)
                cout << ", ";
        }
        cout << ")";
        if (i != count - 1)
            cout << ", ";
    }
    cout << endl;
    displayCount();
}

/*
 Function Name: displayCount
 Parameter: None
 Return Value: None
 display number of the prime pair
 */
void PrimeCalc::displayCount(){
    cout << "COUNT: " << count << endl;
}

/*
 Function Name: checkTypePair
 Parameter: None
 Return Value: None
 check and display the type of the prime pair
 */
void PrimeCalc::checkTypePair(){
    if (this->gap == 2) // twin
        cout << "Twin Prime: " << endl;
    else if (this->gap == 4) // cousin
        cout << "Cousin Prime: " << endl;
    else if (this->gap == 6) // sexy
        cout << "Sexy Prime: " << endl;
    else // others
        cout << "Prime Numbers with " << this->gap << " gap: " << endl;
}

/* Destructor */
PrimeCalc::~PrimeCalc(){
    for (int i = 0; i < pairListSize; ++i){
        delete pairList[i];
        pairList[i] = nullptr;
    }
    delete[] pairList;
    pairList = nullptr;
    
}
